### ISSUE TITLE
Serpent serializer may corrupt data

### ISSUE THREAD

**Issue by shvorin** [libmoon/libmoon#75]
[MoonGen](https://github.com/emmericp/MoonGen) uses Serpent to pass data between it's tasks. The problem is Serpent is unable to serialize some objects. For instance, a queue object (both RX and TX) appears to be broken after deserilization. Also the problem is with MAC address, which is represented by a floating point value in Lua (since Lua does not have 64-bit integers), so the less significant bits may be corrupted after deserialization.

In both issues I used some workaroud in my cloned MoonGen:

-  [queue](https://github.com/QratorLabs/MoonGen/commit/f4565f1a58cd63babdc972322bf110904ae62e39)
- [MAC address](https://github.com/QratorLabs/MoonGen/commit/d5bfe8466597846d621f5256240aac1e12a933fa)

**Comment by emmericp** [comment]
Queues are passed as parameters in most example scripts and seems to work, can you give an example where it is corrupted?

I've seen problems with MAC addresses, yes. But 48 bits should be serializable and I think I once fixed something related to this, I will check this.

**Comment by shvorin** [comment]
Just revert the mentioned commit https://github.com/QratorLabs/MoonGen/commit/f4565f1a58cd63babdc972322bf110904ae62e39 in my MoonGen clone and run
```
./build/MoonGen  ./examples/l3-tcp-syn-ack-flood.lua 0   -c
```
That's about RX queue serialization, so to get an error some traffic must be received at the running machine.
```
...
[INFO]  Device 0 (7C:FE:90:77:D2:92) is up: 40000 MBit/s
rxCount
[FATAL] Lua error in task rxCount
/home/art/MoonGen/build/../libmoon/lua/driver/mlx5.lua:71: bad argument #2 to 'new' (cannot convert 'nil' to 'int')
Stack Traceback
===============
(2) Lua method 'getThroughput' at file '/home/art/MoonGen/build/../libmoon/lua/driver/mlx5.lua:71'
        Local variables:
         self = [Device: id=0]  {initialized:true, rxQueues:table: 0x414071c8, txQueues:table: 0x41f42dc8 (more...)}
...
```
It seems that queue object was corrupted, and so it's dev points to nowhere.

Now I'm trying to reproduce this bug with by task within the pristine repository.

**Comment by shvorin** [comment]
Yes, it also breaks in https://github.com/emmericp/MoonGen repository.
Just grab this file https://github.com/QratorLabs/MoonGen/blob/testing-queue-serialization/examples/l3-tcp-syn-ack-flood.lua and run
```
./build/MoonGen  ./examples/l3-tcp-syn-ack-flood.lua 0   -c
```

**Comment by emmericp** [comment]
Working as expected for me, fresh server, fresh clone:

``
   3  git clone https://github.com/emmericp/MoonGen
    4  cd MoonGen/
    5  ls
    6  ./build.sh 
    7  ./setup-hugetlbfs.sh 
    8  curl https://raw.githubusercontent.com/QratorLabs/MoonGen/testing-queue-serialization/examples/l3-tcp-syn-ack-flood.lua > test.lua
    9  ./build/MoonGen test.lua  0 -c
``

``
[INFO]  Device 0 (68:05:CA:32:44:D8) is up: 40000 MBit/s
rxCount
^C[Device: id=0] RX: 0.00 Mpps, 0 Mbit/s (0 Mbit/s with framing)
[Device: id=0] RX: nan (StdDev 0.00) Mpps, nan (StdDev 0) Mbit/s (nan Mbit/s with framing), total 1 packets with 87 bytes (incl. CRC)
``

**Comment by emmericp** [comment]
i've fixed the queue thing, it was specific to the mlx driver and the way you were using it

mac serialization seems to work for me, can you provide an example where this fails? i've tested it with this test:

```
local mg = require "libmoon"

local function test(mac)
        assert(mg.startTask("thread", mac):wait() == mac)
end

function master()
        test(0xFFFFFFFFFFFF)
        test(0)
        test(0x1234567890AB)
end

function thread(mac)
        return mac
end
```

**Comment by shvorin** [comment]
Thank you for fixing queues, it works.

As for MAC serialization, that's quite old issue, I need some time to restore that circumstance.
